packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/drawable/BackgroundImageDrawable.kt

```diff
@@ -0,0 +1,374 @@
+import com.facebook.react.uimanager.style.BackgroundPosition
+import com.facebook.react.uimanager.style.BackgroundRepeat
+import com.facebook.react.uimanager.style.BackgroundRepeatKeyword
+import com.facebook.react.uimanager.style.BackgroundSize
+import com.facebook.react.uimanager.style.BorderInsets
+import com.facebook.react.uimanager.style.BorderRadiusStyle
+import kotlin.math.ceil
+import kotlin.math.floor
+import kotlin.math.round
+
+internal class BackgroundImageDrawable(
+    private val context: Context,
+    /*
+     * We assume borderRadius & borderInsets to be shared across multiple drawables
+     * therefore we should manually invalidate this drawable when changing either of them
+     */
+    var borderRadius: BorderRadiusStyle? = null,
+    var borderInsets: BorderInsets? = null,
+) : Drawable() {
+  private var needUpdatePath = true
+  private var backgroundImageClipPath: Path? = null
+  private var backgroundPositioningArea: RectF? = null
+  private var backgroundPaintingArea: RectF? = null
+
+  var backgroundImageLayers: List<BackgroundImageLayer>? = null
+    set(value) {
+      if (field != value) {
+        field = value
+        invalidateSelf()
+      }
+    }
+  
+  var backgroundSize: List<BackgroundSize>? = null
+    set(value) {
+      if (field != value) {
+        field = value
+        invalidateSelf()
+      }
+    }
+
+  var backgroundPosition: List<BackgroundPosition>? = null
+    set(value) {
+      if (field != value) {
+        field = value
+        invalidateSelf()
+      }
+    }
+
+  var backgroundRepeat: List<BackgroundRepeat>? = null
+    set(value) {
+      if (field != value) {
+        field = value
+        invalidateSelf()
+      }
+    }
+
+  private val backgroundPaint: Paint =
+      Paint(Paint.ANTI_ALIAS_FLAG).apply {
+        style = Paint.Style.FILL
+      }
+
+  override fun invalidateSelf() {
+    needUpdatePath = true
+    super.invalidateSelf()
+  }
+
+  override fun onBoundsChange(bounds: Rect) {
+    super.onBoundsChange(bounds)
+    needUpdatePath = true
+  }
+
+  override fun setAlpha(alpha: Int) {
+    backgroundPaint.alpha = alpha
+    invalidateSelf()
+  }
+
+  override fun setColorFilter(colorFilter: ColorFilter?) {
+    // do nothing
+  }
+
+  @Deprecated("Deprecated in Java")
+  override fun getOpacity(): Int {
+    val alpha = backgroundPaint.alpha
+    return when (alpha) {
+      255 -> PixelFormat.OPAQUE
+      in 1..254 -> PixelFormat.TRANSLUCENT
+      else -> PixelFormat.TRANSPARENT
+    }
+  }
+
+  override fun draw(canvas: Canvas) {
+    updatePath()
+    canvas.save()
+    if (backgroundImageLayers == null || backgroundImageLayers?.isEmpty() == true) {
+      canvas.restore()
+      return
+    }
+    val backgroundPaintingArea = backgroundPaintingArea ?: return
+    val backgroundPositioningArea = backgroundPositioningArea ?: return
+
+    if (hasInvalidDimensions(backgroundPositioningArea, backgroundPaintingArea)) {
+      canvas.restore()
+      return
+    }
```

Comment:
nit: Same here, maybe just hold off from calling `canvas.save()` until after these early returns?
```suggestion
    if (hasInvalidDimensions(backgroundPositioningArea, backgroundPaintingArea)) {
      return
    }
    
    canvas.save()
```